### PR TITLE
critical css

### PR DESCRIPTION
**Pull request by lenarother**
Hey,

critical-css we used for moccu page (https://github.com/moccu/moccu-website/tree/master/src/moccusite/critical) fas now besome a separate library. Would be greate to get some feedback from you. 

LG,
Magda

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -17,19 +17,26 @@ django-critical-css
    :target: https://travis-ci.org/lenarother/django-critical-css
 
 
-TODO: Explain what your library does in one or two sentences.
+django-critical-css speeds up webpage rendering by saving
+`critical css <http://www.phpied.com/css-and-the-critical-path/>`_ in a database.
```
Comment (schorfES) [inline README.rst]:
```suggestion
 `critical css <https://www.phpied.com/css-and-the-critical-path/>`_ in a database.
```

File: critical/api.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+
+logger = logging.getLogger(__name__)
+
+
+class PenthouseApi(object):
+
+    @property
+    def base_url(self):
+        return 'http://{0}:3000/'.format(settings.PENTHOUSE_HOST)
```
Comment (schorfES) [inline critical/api.py:16]: Do we want to make the port also customizable?

File: critical/api.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                'params: {0}'.format(params))
+
+        return response
+
+    def get_critical_css(self, target_url, target_css):
+        response = self.request_critical_css(target_url, target_css)
+
+        if len(response.text) < 3 or response.text[:3].isdigit():
+            raise PenthouseException(
+                'Invalid pentahouse response ({0}): '
```
Comment (schorfES) [inline critical/api.py]:
```suggestion
                'Invalid penthouse response ({0}): '
```

File: critical/templates/critical/critical.html [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+{% if critical_css %}
+	{% block critical_css %}
+		<style type="text/css">{% block critical_css_inline %}{{ critical_css|safe }}{% endblock %}</style>
+	{% endblock %}
+	{% block preload %}
+		<link rel="preload" as="style" type="text/css" href="{{ css_path }}" onload="this.onload=null;this.rel='stylesheet'" />
+	{% endblock %}
+	{% block noscript %}
+		<noscript><link rel="stylesheet" type="text/css" href="{{ css_path }}?noscript" /></noscript>
```
Comment (schorfES) [inline critical/templates/critical/critical.html]: I think we can drop that `noscript`-get param. I can not remember why I've added this here.

```suggestion
		<noscript><link rel="stylesheet" type="text/css" href="{{ css_path }}" /></noscript>
```

File: docs/usage.rst [hunk 5]
```diff
@@ -1,4 +1,74 @@
 Usage
 =====
 
-TODO: Explain how to use your library.
+Prerequisites
+-------------
+
+To perform its function, django-critical-css requires:
+
+* `penthouse <https://www.npmjs.com/package/penthouse/>`_ service available.
```
Comment (schorfES) [inline docs/usage.rst:9]: django-critical-css requires the service which we've build around the _penthouse_ package.

For that we need to do the following things:

1. Make https://github.com/moccu/penthouse-service repo public (+ update documentation?)
2. Publish that via NPM (+ update documentation)
3. Refer to docker repo: https://cloud.docker.com/u/moccu/repository/docker/moccu/penthouse-service

File: docs/usage.rst [hunk 6]
```diff
@@ -1,4 +1,74 @@
...
+-------------------------------------
+
+django-critical-css provides the `critical_css` templatetag to inline critical css
+from the database into your template using ``templates/critical/critical.html``.
+
+.. code-block:: text
+
+    {% load critical_tags %}
+
+      {% critical_css 'your/css/styles_path.css' %}
```
Comment (schorfES) [inline docs/usage.rst]: Bring `{% load critical_tags %}` and `{% critical_css ... %}` to same indentation:

```suggestion
    {% critical_css 'your/css/styles_path.css' %}
```

File: docs/usage.rst [hunk 7]
```diff
@@ -1,4 +1,74 @@
...
+
+If page does not have any critical css saved in the database yet or the css path of the saved
+critical-cms-object is different than the argument of `critical_css` the tag will return a normal css link.
```
Comment (schorfES) [inline docs/usage.rst]:
```suggestion
 critical-cms-object is different than the argument of `critical_css` the tag will return a normal stylesheet link-tag.
```

File: docs/usage.rst [hunk 8]
```diff
@@ -1,4 +1,74 @@
...
+
+
+Additional settings
+-------------------
+
+* By setting ``CRITICAL_CSS_ACTIVE`` to `False` you can deactivate css calculation
+  for your dev environment. By default critical css calculation is always active.
+* ``PENTHOUSE_CONFIG`` is a dict with additional configuration options that will
+  be sent to the penthouse service along with the page url and css path. See available
+	options `here <http://www.phpied.com/css-and-the-critical-path/>`_. By default
```
Comment (schorfES) [inline docs/usage.rst]:
```suggestion
 	options `here <https://github.com/moccu/penthouse-service#options>`_. By default
```

File: examples/critical-example/README.rst [hunk 9]
```diff
@@ -0,0 +1,62 @@
+Run Example
+===========
+
+1. Run penthouse container on port 3000.
+   Check whether it is running:
+
+.. code-block:: text
+
+    http://0.0.0.0:3000/?url=https://www.moccu.com/&css=https://www.moccu.com/static/css/styles.css
```
Comment (schorfES) [inline examples/critical-example/README.rst:9]: Do you want to use moccu-urls for the example call?

File: examples/critical-example/criticalexample/templates/example.html [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    </header>
+    <div class="github-fork-ribbon-wrapper right">
+      <div class="github-fork-ribbon">
+        <a href="https://github.com/tonystar/bootstrap-template">Fork me on GitHub</a>
+      </div>
+    </div>
+    <script src="//cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
+    <script src="//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.5/js/bootstrap.min.js"></script>
+    <script src="//cdn.rawgit.com/tonystar/bootstrap-plugins/master/dist/js/bootstrap-plugins.min.js"/>
+    <!--script src="custom.js"></script-->
```
Comment (schorfES) [inline examples/critical-example/criticalexample/templates/example.html]: The example don't use any custom scripts and its commented, so you can drop this line.

**Review by schorfES — COMMENTED** [review]
I've added some comments to the code. 

I have one general question: The logging seems to be always enabled which would result in some _noise_. Should an extra configuration like `CRITICAL_CSS_VERBOSE = True` to be added?

File: examples/critical-example/README.rst (hunk 9 above)
Comment (lenarother) [reply examples/critical-example/README.rst:9]: Any other ideas for working example?

**Comment by lenarother** [comment]
Marged in another PR

**Pull request closed without merging**